// ipfs-block-service/lib/index.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var parallelLimit = require('run-parallel-limit');
var pull = require('pull-stream');

// BlockService is a hybrid block datastore. It stores data in a local
// datastore and may retrieve data from a remote Exchange.
// It uses an internal `datastore.Datastore` instance to store values.
module.exports = function () {
  function BlockService(ipfsRepo) {
    _classCallCheck(this, BlockService);

    this._repo = ipfsRepo;
    this._bitswap = null;
  }

  _createClass(BlockService, [{
    key: 'goOnline',
    value: function goOnline(bitswap) {
      this._bitswap = bitswap;
    }
  }, {
    key: 'goOffline',
    value: function goOffline() {
      this._bitswap = null;
    }
  }, {
    key: 'isOnline',
    value: function isOnline() {
      return this._bitswap != null;
    }

    // Note: we have to pass the CID, so that bitswap can then use it for
    // the smart selectors. For now, passing the CID is used so that we know
    // the right multihash, this means that now we have multiple hashes official
    // support \o/

  }, {
    key: 'put',
    value: function put(blockAndCID, callback) {
      callback = callback || function () {};
      if (!blockAndCID) {
        return callback(new Error('Missing block and CID'));
      }

      pull(pull.values([blockAndCID]), this.putStream(), pull.onEnd(callback));
    }
  }, {
    key: 'putStream',
    value: function putStream() {
      var ps = void 0;
      if (this.isOnline()) {
        // NOTE: This will have to change in order for bitswap
        // to understand CID
        ps = this._bitswap.putStream();
      } else {
        ps = this._repo.blockstore.putStream();
      }

      return pull(pull.map(function (blockAndCID) {
        return {
          data: blockAndCID.block.data,
          key: blockAndCID.cid.multihash
        };
      }), ps);
    }
  }, {
    key: 'get',
    value: function get(cid, callback) {
      pull(this.getStream(cid), pull.collect(function (err, result) {
        if (err) {
          return callback(err);
        }
        callback(null, result[0]);
      }));
    }
  }, {
    key: 'getStream',
    value: function getStream(cid) {
      if (this.isOnline()) {
        return this._bitswap.getStream(cid.multihash);
      }

      return this._repo.blockstore.getStream(cid.multihash);
    }
  }, {
    key: 'delete',
    value: function _delete(cids, callback) {
      var _this = this;

      if (!Array.isArray(cids)) {
        cids = [cids];
      }

      parallelLimit(cids.map(function (cid) {
        return function (next) {
          _this._repo.blockstore.delete(cid.multihash, next);
        };
      }), 100, callback);
    }
  }]);

  return BlockService;
}();
